File: clients/v2boxkit_client/lib/src/services/subscription_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../model/models.dart';

class SubscriptionFetchResult {
  const SubscriptionFetchResult({
    required this.notModified,
    this.body,
    this.etag,
    this.lastModified,
  });

  final bool notModified;
  final String? body;
  final String? etag;
  final String? lastModified;
}

class SubscriptionService {
  SubscriptionService({http.Client? client})
    : _client = client ?? http.Client();

  static const _maximumBytes = 4 * 1024 * 1024;
  final http.Client _client;

  Future<SubscriptionFetchResult> fetch(ProxySubscription subscription) async {
    if (subscription.url.scheme != 'https') {
      throw const FormatException('订阅地址必须使用 HTTPS');
    }
    final headers = <String, String>{
      'accept': 'text/plain, application/json, application/yaml, */*',
      'user-agent': 'V2BoxKit/1.0',
    };
    if (subscription.etag case final etag?) {
      headers['if-none-match'] = etag;
    }
    if (subscription.lastModified case final modified?) {
      headers['if-modified-since'] = modified;
    }

    final response = await _client
        .get(subscription.url, headers: headers)
        .timeout(const Duration(seconds: 15));
    if (response.statusCode == 304) {
      return SubscriptionFetchResult(
        notModified: true,
        etag: response.headers['etag'] ?? subscription.etag,
        lastModified:
            response.headers['last-modified'] ?? subscription.lastModified,
      );
    }
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw http.ClientException(
        '订阅服务器返回 HTTP ${response.statusCode}',
        subscription.url,
      );
    }
    if (response.bodyBytes.length > _maximumBytes) {
      throw const FormatException('订阅内容超过 4 MiB 限制');
    }
    return SubscriptionFetchResult(
      notModified: false,
      body: utf8.decode(response.bodyBytes, allowMalformed: false),
      etag: response.headers['etag'],
      lastModified: response.headers['last-modified'],
    );
  }

  void close() => _client.close();
}

File: Sources/App/Services/SubscriptionService.swift
import Foundation

struct SubscriptionFetchResult: Sendable {
  let nodes: [ProxyNode]?
  let etag: String?
  let lastModified: String?

  var isNotModified: Bool { nodes == nil }
}

struct SubscriptionService {
  enum SubscriptionError: LocalizedError {
    case invalidURL
    case invalidResponse
    case unsupportedEncoding

    var errorDescription: String? {
      switch self {
      case .invalidURL: "订阅地址必须是 HTTPS URL"
      case .invalidResponse: "订阅服务器返回失败"
      case .unsupportedEncoding: "订阅内容不是 UTF-8 文本"
      }
    }
  }

  func fetchNodes(from url: URL) async throws -> [ProxyNode] {
    let subscription = ProxySubscription(name: url.host ?? "订阅", url: url)
    return try await fetch(subscription: subscription).nodes ?? []
  }

  func fetch(subscription: ProxySubscription) async throws -> SubscriptionFetchResult {
    guard subscription.url.scheme?.lowercased() == "https" else {
      throw SubscriptionError.invalidURL
    }

    var request = URLRequest(url: subscription.url)
    request.timeoutInterval = 20
    request.setValue("V2BoxKit/1.0", forHTTPHeaderField: "User-Agent")
    if let etag = subscription.etag {
      request.setValue(etag, forHTTPHeaderField: "If-None-Match")
    }
    if let lastModified = subscription.lastModified {
      request.setValue(lastModified, forHTTPHeaderField: "If-Modified-Since")
    }
    let (data, response) = try await URLSession.shared.data(for: request)

    guard let httpResponse = response as? HTTPURLResponse else {
      throw SubscriptionError.invalidResponse
    }
    if httpResponse.statusCode == 304 {
      return SubscriptionFetchResult(
        nodes: nil,
        etag: httpResponse.value(forHTTPHeaderField: "ETag") ?? subscription.etag,
        lastModified: httpResponse.value(forHTTPHeaderField: "Last-Modified")
          ?? subscription.lastModified
      )
    }
    guard (200..<300).contains(httpResponse.statusCode) else {
      throw SubscriptionError.invalidResponse
    }
    guard let text = String(data: data, encoding: .utf8) else {
      throw SubscriptionError.unsupportedEncoding
    }
    return SubscriptionFetchResult(
      nodes: try ShareLinkParser.parseMany(text),
      etag: httpResponse.value(forHTTPHeaderField: "ETag"),
      lastModified: httpResponse.value(forHTTPHeaderField: "Last-Modified")
    )
  }
}
